### ISSUE TITLE
Only generate effective POMs on demand

### ISSUE THREAD

**Issue by anthonyvdotbe** [microsoft/vscode-maven#579]
**Describe the bug**

I launch VS Code with a workspace that contains a lot of `pom.xml` files (though they are all several levels deep. There's no top-level folder with a `pom.xml` file), but without any intention to work on any Java project. The `Language Support for Java` language server doesn't start, and rightly "understands" that I'm not working on any Java project.
However, for nearly 30(!) minutes after opening VS Code, the extension is busy generating the effective POMs for each and every `pom.xml` it finds. What's worse, is that I have not modified any POM for at least a couple days. Then after it's finally done after half an hour, I close VS Code, launch the same workspace... and it starts all over again...
Note that this is very similar to #319. In general, generation of effective POMs happens way too eagerly & results in an enormous waste of resources.

**To Reproduce**
1. have a workspace that contains a lot of `pom.xml` files, e.g. this repo alone has around 50 already: https://github.com/eclipse-ee4j/jersey/
2. open Process Explorer (see screenshot below) and just watch the processes get launched one after the other endlessly

**Expected behavior**
As long as I don't open a `pom.xml` file or a Java file, or there's any other indication that I might work on some project, don't generate the effective POM.
It's much better to sometimes be too lazy, than to constantly be too eager.
Also, at the very least there should be an indication that this is happening, so users are aware of all the activity and can decide to disable the extension by default, and only enable it when they really need it.

**Environments (please complete the following information as much as possible):**
 - OS: Windows 10
 - VS Code version: 1.53.2 (user setup) (622cb03f7e070a9670c94bae1a45d78d7181fbd4)
 - Extension version: 0.27.1

**Screenshots**
![image](https://user-images.githubusercontent.com/15714253/109856065-57afba80-7c59-11eb-8dbf-0cbf6e93175d.png)

**Additional context**
The processes are all of the form:
```
C:\Windows\system32\cmd.exe /d /s /c "mvn help:effective-pom -Doutput="c:\Users\...\AppData\Roaming\Code\User\workspaceStorage\8075741f9363c45b8996f0027c30d3e0\vscjava.vscode-maven\2c1858093c005cebb9bd50e7187e96cb" -f "c:\Users\...\com.github\...\pom.xml""
```

**Comment by Eskibear** [comment]
Sorry about the bad experience. Here is the story.

In Maven explorer, we show plugins and plugin-specific goal. These information is not available before we calculate effective pom of a project (let me know if there's cheap way to achieve this). 

In the initial version, we did calculate effective pom on demand. Only when you expand "Plugins" node, the extension starts the calculation. But since the calculation takes time, users were complaining about slowness. That's why we execute the process in background on startup, to fetch the information earlier, silently.

For each project we have a corresponding task. For workspaces containing large number of pom files, we had issue #319 . And to mitigate it, we made a queue to execute task one by one in order not to occupy all the CPU cores. The side effect is, the whole process costs a long period of time, and that's why your "30 min" story happened.



The ideal solution is to find a cheap and effective way to get those information, instead of calling maven-help-plugin.

A mitigation is as you mentioned, becoming lazy. For security and transparency, even prompt for user's confirmation when it has to create a background process. 

If you have further thoughts, please share here. Thanks.

**Comment by anthonyvdotbe** [comment]
Thanks, I understand it's hard to keep all users happy.

Here's my thoughts on this:
* add a boolean setting to configure whether the `Plugins` nodes should be populated on demand. I barely ever use this functionality
* implement a basic cache (of either the effective POMs, or just the list of plugins). Then on startup, immediately list all `pom.xml` files & use the cache to populate the `Plugins` nodes. Then update the cache afterwards (using just file timestamp & size is sufficient to detect changes. If ever it wouldn't, the user can simply refresh manually).
* another idea is to find a middle ground between `on demand` and `ASAP`. For example, when I open a file (on startup, this should also check the files that are still open from last time), look for any `pom.xml` in its current & parent directories, and load the `Plugins` sections of those projects: I believe it's uncommon for a user to dive straightaway into the `Plugins` node, without first having opened a file in the project

I believe there should at least be a setting to have everything done on demand (and it would be good to either set its default value to `true`, or show a notification to explicitly ask the user what they want).

**Comment by Eskibear** [comment]
I agree that a cache is needed, no matter whether it's populated ASAP or on demand. Then at least after reloading window, you don't have to re-caclulate the whole bunch of things from beginning. I'll start with this one. 

About the boolean setting you mentioned, I'm thinking of its default value. If I remember correctly, we did caclulate effective on demand, explicitly when "Plugins" node was expanded. At that time, users were complaining that it was extremly slow to see the list of plugins, and then we chose to pre-caclulate it on startup. 
Notification is a safer way to go, that we let users decide by themselves. But we are trying to add as fewer notifications as possible, as they have already been abused in vscode. When you open a folder, you see so many notifications on the screen awaiting for your choice, and we don't think that's a good experience.

**Comment by anthonyvdotbe** [comment]
About the notification: I meant to only show a notification once, when the user first installs/updates to the version with the new setting.

**Comment by Eskibear** [comment]
Yes, I understand your point. We can detect whether the setting is explicitly set, otherwise we show a notification.

Even it only shows once on startup, together similar notifications provided by other extensions, it's also possible to provide bad onboarding experience. Instead of showing the notification on startup, I think maybe it would be better at the first time when user explicitly expand "Plugins" node or any other operation requiring calculation of effective pom... 

And for the scope of the setting, it would be much easier to make it global. It's more like a user's preference, and I don't see much sense to make it workspace level for the moment.

**Comment by anthonyvdotbe** [comment]
> I think maybe it would be better at the first time when user explicitly expand "Plugins" node or any other operation requiring calculation of effective pom

Yes, that'd be great (assuming the setting is "on demand" by default).

> And for the scope of the setting, it would be much easier to make it global. It's more like a user's preference, and I don't see much sense to make it workspace level for the moment.

Yes, I agree.

**Comment by Eskibear** [comment]
Effective POM will be no longer auto calculated on startup. And cache based on last modified timestamp is added, to avoid re-calculation.

**Comment by Eskibear** [comment]
shipped in v0.29.0
